From 8dfead68c20251fccb39113ee864fbbafc5b7f76 Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@nbd.name>
Date: Mon, 14 Oct 2024 19:11:10 +0200
Subject: [PATCH] wifi: mt76: mt7915: hold dev->mutex while interacting with
 the thermal state

Avoid issuing MCU calls during a hardware restart

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---
 mt7915/init.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- a/mt7915/init.c
+++ b/mt7915/init.c
@@ -52,7 +52,9 @@ static ssize_t mt7915_thermal_temp_show(
 
 	switch (i) {
 	case 0:
+		mutex_lock(&phy->dev->mt76.mutex);
 		temperature = mt7915_mcu_get_temperature(phy);
+		mutex_unlock(&phy->dev->mt76.mutex);
 		if (temperature < 0)
 			return temperature;
 		/* display in millidegree celcius */
@@ -94,9 +96,8 @@ static ssize_t mt7915_thermal_temp_store
 	}
 
 	phy->throttle_temp[i - 1] = val;
-	mutex_unlock(&phy->dev->mt76.mutex);
-
 	ret = mt7915_mcu_set_thermal_protect(phy);
+	mutex_unlock(&phy->dev->mt76.mutex);
 	if (ret)
 		return ret;
 
@@ -158,7 +159,9 @@ mt7915_thermal_set_cur_throttle_state(st
 	 * cooling_device convention: 0 = no cooling, more = more cooling
 	 * mcu convention: 1 = max cooling, more = less cooling
 	 */
+	mutex_lock(&phy->dev->mt76.mutex);
 	ret = mt7915_mcu_set_thermal_throttling(phy, throttling);
+	mutex_unlock(&phy->dev->mt76.mutex);
 	if (ret)
 		return ret;
 
